makefile

```diff
@@ -2,121 +2,125 @@ DEBUG_VERSION := 1603
+ifneq ($(findstring withInlineAsm=1,$(CFLAGS)),)
+	EXTRAOBJS += inline-asm.o
+	CFLAGS += $(shell llvm-config --cflags --system-libs --libs core orcjit native)
+	CXXFLAGS += $(shell llvm-config --cxxflags --system-libs --libs core orcjit native) -fexceptions
 endif
 
-ifneq ($(findstring withInjectNative=1,${CFLAGS}),)
-OBJS := ${OBJS} inject.o tcc/libtcc1.a tcc/libtcc.a
+ifneq ($(findstring withInjectNative=1,$(CFLAGS)),)
+	OBJS += inject.o tcc/libtcc1.a tcc/libtcc.a
 endif
 
-COBJS = ${OBJS} compiler.o
+COBJS = $(OBJS) compiler.o
 
-# .SUFFIXES:	.o .c
+# FIXME: maybe a bit hackish?
+OBJS := $(foreach obj,$(OBJS),$(BUILDDIR)/$(obj))
+COBJS := $(foreach obj,$(COBJS),$(BUILDDIR)/$(obj))
+EXTRAOBJS := $(foreach obj,$(EXTRAOBJS),$(BUILDDIR)/$(obj))
 
-all: CFLAGS := $(CFLAGS) -O2
-all: cxx
-all: ctr ctrconfig
+all: CFLAGS += -O2
+all: remove_libsocket_build cxx ctr ctrconfig
+
+remove_libsocket_build:
+	rm -rf src/lib/libsocket/libsocket/build/
 
 ctrconfig:
-	$(CC) ctrconfig.c -o ctrconfig
-
-debug: CFLAGS := ${CFLAGS} -DDEBUG_BUILD -DDEBUG_BUILD_VERSION=${DEBUG_BUILD_VERSION} -Og -g3 -ggdb3 -Wno-unused-function
-debug: cxx
-debug: ctr
-debug:
-	sed -i -e "1s/.*/${new_makefile_l1}/" makefile
-
-install:
-	echo -e "install directly from source not allowed.\nUse citron_autohell instead for installs"
-	exit 1;
-ctr:	$(OBJS) $(EXTRAOBJS)
-	$(CXX) -fopenmp $(EXTRAOBJS) $(OBJS) ${CXXFLAGS}  -rdynamic -lm -ldl -lbsd -lpcre -lffi -lpthread ${LEXTRACF} -o ctr
-
-libctr: CFLAGS := $(CFLAGS) -fPIC -DCITRON_LIBRARY
-libctr: symbol_cxx
-libctr: $(OBJS)
-	$(CC) $(OBJS) -shared -export-dynamic -ldl -lbsd -lpcre -lffi -lpthread -o libctr.so
-
-compiler: CFLAGS := $(CFLAGS) -D comp=1
-compiler: cxx
-compiler: $(COBJS)
-	$(CC) $(COBJS) -rdynamic -lm -ldl -lbsd -lpcre -lffi -lprofiler -lpthread ${LEXTRACF} -o ctrc
+	$(CC) src/ctrconfig.c -o $(BUILDDIR)/ctrconfig
 
-cxx:
-	echo "blah"
+$(BUILDDIR):
+	mkdir -p $@
+
+$(BUILDDIR)/ctr: $(BUILDDIR) $(OBJS) $(EXTRAOBJS)
+	$(CXX) -fopenmp $(EXTRAOBJS) $(OBJS) $(CXXFLAGS) -rdynamic -lm -ldl -lbsd -lpcre -lffi -lpthread $(LEXTRACF) -o $@
+
+ctr: $(BUILDDIR)/ctr
+
+$(BUILDDIR)/libctr.so: CFLAGS += -fPIC -DCITRON_LIBRARY
+$(BUILDDIR)/libctr.so: $(OBJS)
+	$(CC) $< -shared -export-dynamic -ldl -lbsd -lpcre -lffi -lpthread -o $@
 
-tcc/%.a:
-	cd tcc && ./configure --prefix=$(realpath build) && $(MAKE) $<
+libctr: $(BUILDDIR)/libctr.so
 
-# %.o: %.c
-# 	echo "$<"
-# 	$(CC) $(CFLAHS) -c $< -o $@ >/dev/null 2>&1
+# compiler: CFLAGS := $(CFLAGS) -D comp=1
+# compiler: cxx
+# compiler: $(COBJS)
+# 	cd $(BUILDDIR)
+# 	$(CC) $< -rdynamic -lm -ldl -lbsd -lpcre -lffi -lprofiler -lpthread $(LEXTRACF) -o ctrc
+# 	cd -
 
-libsocket.so:
-	make -C libsocket
-	cp libsocket/libsocket.so libsocket.so
+install: $(BUILDDIR)/ctr $(BUILDDIR)/libctr.so $(BUILDDIR)/ctrconfig
+	sudo install $(BUILDDIR)/ctr $(DESTDIR)
+	sudo install $(BUILDDIR)/libctr.so $(LIBDIR)
+	sudo install $(BUILDDIR)/ctrconfig $(DESTDIR)
```

Comment:
Let's not use `sudo` here, the user can run `make` as root if they wanna install to places that require perms